cloud/utils/retry.go

```diff
@@ -0,0 +1,63 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package utils
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+// RetryOnFailure allows to retry a lambda up to a failure condition depending on the implementation
+type RetryOnFailure interface {
+	Run() error
+}
+
+// exponentialBackOffRetrier implements RetryOnFailure with an exponential back off on retrying funcToRetry
+type exponentialBackOffRetrier struct {
+	funcToRetry func() error
+	backoff     backoff.BackOff
+}
+
+// NewRetrier returns a retrier using an exponential backoff to retry funcToRetry.
+func NewRetrier(funcToRetry func() error) RetryOnFailure {
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.InitialInterval = 500 * time.Millisecond
+	expBackoff.RandomizationFactor = 0.5
+	expBackoff.Multiplier = 1.5
+	expBackoff.MaxElapsedTime = 3 * time.Second
+
+	return &exponentialBackOffRetrier{
+		funcToRetry: funcToRetry,
+		backoff:     expBackoff,
+	}
+}
+
+// Run will run r.funcToRetry and retry on failure with an ExponentialBackOff up to r.backoff.MaxElapsedTime
+func (r *exponentialBackOffRetrier) Run() error {
```

Comment:
backoff has already builtin func `backoff.Retry` i think it will let us avoid the interface and the run function

Reply:
The reason I did this was that at first I wanted to filter on a given error type. I removed this afterward. I think like that it's nice because it allows us to have a single "New" function and avoid having to re-set the parameters everytime.

Reply:
sure, but we can use the params as constant config :thinking_face: ?   I mean if we gonna have run, then we don't have to use backoff lib and instead our own retry but the point of suggesting [github.com/cenkalti/backoff](http://github.com/cenkalti/backoff) it's neat in it's config and. retry mechanism
with the current Run()  i think we don't benefit from that!

Reply:
resolved over slack